cpp/include/raft/sparse/distance/detail/utils.cuh

```diff
@@ -37,6 +41,199 @@ inline int max_cols_per_block()
+                     raft::distance::DistanceType metric)
+{
+  if (nnz == 0) return;
+  auto stream = resource::get_cuda_stream(handle);
+
+  int blocks = raft::ceildiv<size_t>((size_t)nnz, tpb);
+  if (metric == raft::distance::DistanceType::L2Expanded) {
+    epilogue_on_csr_kernel<<<blocks, tpb, 0, stream>>>(
+      compressed_C,
+      rows,
+      cols,
+      Q_sq_norms,
+      R_sq_norms,
+      nnz,
+      [] __device__ __host__(value_t dot, value_t q_norm, value_t r_norm) -> value_t {
+        return value_t(-2.0) * dot + q_norm + r_norm;
+      });
+  } else if (metric == raft::distance::DistanceType::L2SqrtExpanded) {
+    epilogue_on_csr_kernel<<<blocks, tpb, 0, stream>>>(
+      compressed_C,
+      rows,
+      cols,
+      Q_sq_norms,
+      R_sq_norms,
+      nnz,
+      [] __device__ __host__(value_t dot, value_t q_norm, value_t r_norm) -> value_t {
+        return raft::sqrt(value_t(-2.0) * dot + q_norm + r_norm);
+      });
+  } else if (metric == raft::distance::DistanceType::CosineExpanded) {
+    epilogue_on_csr_kernel<<<blocks, tpb, 0, stream>>>(
+      compressed_C,
+      rows,
+      cols,
+      Q_sq_norms,
+      R_sq_norms,
+      nnz,
+      [] __device__ __host__(value_t dot, value_t q_norm, value_t r_norm) -> value_t {
+        return value_t(1.0) - dot / (q_norm * r_norm);
+      });
+  }
+  RAFT_CUDA_TRY(cudaPeekAtLastError());
+}
+
+template <typename value_idx, typename value_t>
+__global__ void faster_dot_on_csr_kernel(value_t* __restrict__ dot,
+                                         const value_idx* __restrict__ indptr,
+                                         const value_idx* __restrict__ cols,
+                                         const value_t* __restrict__ A,
+                                         const value_t* __restrict__ B,
+                                         const value_idx nnz,
+                                         const value_idx n_rows,
+                                         const value_idx dim)
+{
+  auto vec_id  = threadIdx.x;
+  auto lane_id = threadIdx.x & 0x1f;
+
+  extern __shared__ char smem[];
+  value_t* s_A      = (value_t*)smem;
+  value_idx cur_row = -1;
+
+#pragma unroll
+  for (int row = blockIdx.x; row < n_rows; row += gridDim.x) {
+#pragma unroll
+    for (int dot_id = blockIdx.y + indptr[row]; dot_id < indptr[row + 1]; dot_id += gridDim.y) {
+      if (dot_id >= nnz) { return; }
+      const value_idx col               = cols[dot_id] * dim;
+      const value_t* __restrict__ B_col = B + col;
+
+      if (threadIdx.x == 0) { dot[dot_id] = 0.0; }
+      __syncthreads();
+
+      if (cur_row != row) {
+#pragma unroll
+        for (value_idx k = vec_id; k < dim; k += blockDim.x) {
+          s_A[k] = A[row * dim + k];
+        }
+        cur_row = row;
+      }
+
+      value_t l_dot_ = 0.0;
+#pragma unroll
+      for (value_idx k = vec_id; k < dim; k += blockDim.x) {
+        asm("prefetch.global.L2 [%0];" ::"l"(B_col + k + blockDim.x));
```

Comment:
Hi @achirkin, I guess this symbol checking it concerns here, what should I do to keep the same function (this is very useful for better performance.)